src/Webapi/Canvas/Webapi__Canvas__Canvas2d.re

```diff
@@ -78,131 +78,147 @@ type style(_) =
   | Pattern: style(pattern);
 
 /* 2d Canvas API, following https://simon.html5.org/dump/html5-canvas-cheat-sheet.html */
-[@bs.send.pipe : t] external save : unit = "";
-[@bs.send.pipe : t] external restore : unit = "";
+[@send] external save: t => unit;
+[@send] external restore: t => unit;
 
 /* Transformation */
-[@bs.send.pipe : t] external scale : (~x: float, ~y: float) => unit = "";
-[@bs.send.pipe : t] external rotate : float => unit = "";
-[@bs.send.pipe : t] external translate : (~x: float, ~y: float) => unit = "";
-[@bs.send.pipe : t] external transform : (~m11: float, ~m12: float, ~m21: float, ~m22: float, ~dx: float, ~dy: float) => unit = "";
-[@bs.send.pipe : t] external setTransform : (~m11: float, ~m12: float, ~m21: float, ~m22: float, ~dx: float, ~dy: float) => unit = "";
+[@send] external scale: (t, ~x: float, ~y: float) => unit;
+[@send] external rotate: (t, float) => unit;
+[@send] external translate: (t, ~x: float, ~y: float) => unit;
+[@send]
+external transform:
+  (t, ~m11: float, ~m12: float, ~m21: float, ~m22: float, ~dx: float, ~dy: float) => unit;
+[@send]
+external setTransform:
+  (t, ~m11: float, ~m12: float, ~m21: float, ~m22: float, ~dx: float, ~dy: float) => unit;
 
 /* Compositing */
-[@bs.set] external globalAlpha : (t, float) => unit = "";
-[@bs.set] external globalCompositeOperation : (t, Composite.t) => unit = "";
+[@set] external globalAlpha: (t, float) => unit;
+[@set] external globalCompositeOperation: (t, Composite.t) => unit;
 
 /* Line Styles */
-[@bs.set] external lineWidth : (t, float) => unit = "";
-[@bs.set] external lineCap : (t, LineCap.t) => unit = "";
-[@bs.set] external lineJoin : (t, LineJoin.t) => unit = "";
-[@bs.set] external miterLimit : (t, float) => unit = "";
+[@set] external lineWidth: (t, float) => unit;
+[@set] external lineCap: (t, LineCap.t) => unit;
+[@set] external lineJoin: (t, LineJoin.t) => unit;
+[@set] external miterLimit: (t, float) => unit;
 
 /* Colors, Styles, and Shadows */
-[@bs.set] external setFillStyle : (t, 'a) => unit = "fillStyle";
-[@bs.set] external setStrokeStyle : (t, 'a) => unit = "strokeStyle";
+[@set] external setFillStyle: (t, [@ignore] style('a), 'a) => unit = "fillStyle";
```

Comment:
These could also just be separate functions and still do the same thing at zero cost without fancy phantom types. Less cool but simpler. Wouldn't scale if there where lots and lots of these functions and type variants though.
```
[@set] external setColorFillStyle: (t, string) => unit = "fillStyle";
[@set] external setGradientFillStyle: (t, gradient) => unit = "fillStyle";
[@set] external setPatternFillStyle: (t, pattern) => unit = "fillStyle";
```

Reply:
Simpler to maintain, not simpler to use. I think I really like this approach as a way to map TS-style `string | gradient | pattern` types and I might look at other places to use it :)